File: project/function_trace/launch_function.py
import os
import sys
import time
import subprocess

def get_pid():
    pid = os.getpid()
    return pid

# fibbonacci function
def fib(n):
    if n < 2:
        return n
    else:
        return fib(n-1) + fib(n-2)


def launch_function(num):
    current_pid = get_pid()
    command = f"echo {current_pid} > /sys/fs/ghost/enclave_1/tasks"
    print(command)
    subprocess.call(command, shell=True)

    fib(num)

    print("Finished!")


if __name__ == "__main__":
    num = int(sys.argv[1])
    launch_function(num)

File: project/function_trace/launch_function.cc
#include <iostream>
#include <unistd.h>
#include <cstdlib>

unsigned long long fibonacci(int n) {
    if (n <= 1) {
        return 1;
    }
    return fibonacci(n - 1) + fibonacci(n - 2);
}

int main(int argc, char *argv[]) {

    // get the pid of the current process
    std::string pid = std::to_string(getpid());
    // Add the process to enclave_1
    std::string command = "echo " + pid + " > /sys/fs/ghost/enclave_1/tasks";

    int ret = std::system(command.c_str());
    if (ret == 0) {
        // if successful, print the command
        std::cout << command << std::endl;
    } else {
        // if failed, print the error message
        std::cout << "Failed to add to enclave_1" << std::endl;
    }

    int arg = atoi(argv[1]);
    unsigned long long n = fibonacci(arg);
    std::cout << n << std::endl;
    return 0;
}